R/get_ds_rt.R

```diff
@@ -183,36 +225,57 @@ get_ds_rt <- function(counts, popn, std_popn, scale = NULL, power = NULL,
     stop("both `dist` and `interval` must be supplied to construct confidence intervals")
   }
 
-  # check if confidence intervals should be constructed
   if (!is.null(dist) & !is.null(interval)) {
     if (clean_strata == "zero") {
       stop("construction of confidence interval with clean_strata = 'zero' not yet implemented")
     }
-    if (stringr::str_detect(dist, stringr::regex("^normal$", ignore_case = TRUE))) { # normal distribution
-      # calculate variance of directly standardized rate
-      var <- get_ds_rt_var(df)
-      threshold <- (1 - 0.997) / 2
-      check_value <- stats::pnorm(0, mean = df_dsr$dsr, sd = sqrt(var))
-      if (!is.na(check_value)) {
-        if (check_value > threshold) { # check if normal distribution should be avoided
-          warning("more than 0.15% of the probability mass lies below 0 for the estimate, consider using a different probability distribution")
-        }
+
+    # check validity of distribution name calculate confidence interval
+    dist_name <- get_dist_name(dist)
+    var <- get_ds_rt_var(df)
+
+    if (dist_name %in% c("normal", "lognormal")) {
+      if (!is.null(method)) {
+        stop("`method` must be NULL when `dist` is 'normal' or 'lognormal'")
+      }
+    }
+
+    if (dist_name == "normal") {
+      df_dsr <- df_dsr %>%
+        dplyr::mutate(
+          get_ci_norm(interval = interval, estimate = .data$dsr, variance = var)
+        )
+    } else if (dist_name == "lognormal") {
+      df_dsr <- df_dsr %>%
+        dplyr::mutate(
+          var_log = (1 / (.data$dsr**2)) * var, # estimation found in Direct Standardization section of STDRATE Procedure Details
+          get_ci_lnorm(interval = interval, estimate = .data$dsr, variance_log = .data$var_log)
+        )
+    } else if (dist_name == "gamma") {
+      weights <- df %>%
+        dplyr::mutate(w_j = (std_popn / sum(std_popn)) * (1 / popn)) %>%
+        dplyr::pull(.data$w_j) %>%
+        list()
+
+      # if user doesn't specify a method, use the default
+      if (is.null(method)) {
+        method <- "tcz06"
       }
-    } else if (stringr::str_detect(dist, stringr::regex("^log", ignore_case = TRUE))) { # log normal distribution
-      # calculate variance of directly standardized rate
-      var <- get_ds_rt_var(df)
+
+      df_dsr <- df_dsr %>%
+        dplyr::mutate(get_ci_gamma(
+          interval = interval, estimate = .data$dsr, weights = weights,
+          variance = var, method = method
+        ))
     } else {
-      stop("this function can only construct confidence intervals using normal and log normal distributions")
+      stop("`dist` should be one of 'normal', 'log normal', or 'gamma'")
```

Comment:
clarify that "for a directly standardized rate, `dist` should be one of..." (to remind user that this is coming from their attempt to get a DSR)